### PR TITLE
Initiate pricing strategy inside the search index class in order to make it easier to override.

### LINKED ISSUE #1507: search_index strategy should not be instantiated outside of a class
The Strategy selector is instantiated outside of a class in the search_index.py.  This makes it difficult to use a custom strategy (with different arguments to the method) as we cannot override this functionality in the oscar base class

### PR DESCRIPTION
**Pull request by sasha0**
Fixes #1507.

### CODE REVIEW

**Comment by mvantellingen** [comment]
This seems to initiate the strategy multiple times which can have a negative performance impact

**Comment by solarissmoke** [comment]
> This seems to initiate the strategy multiple times which can have a negative performance impact.

Easiest way to address this would be to locally cache the result of `ProductIndex.get_strategy()` and return that.

**Comment by sasha0** [comment]
Add simple caching. WDYT?

File: src/oscar/apps/search/search_indexes.py [hunk 1]
```diff
@@ -57,7 +56,13 @@ def prepare_rating(self, obj):
     # most common case is for customers to see the same prices and stock levels
     # and so we implement that case here.
 
+    def get_strategy(self):
+        if not self._strategy:
```
Comment (solarissmoke) [inline src/oscar/apps/search/search_indexes.py]: Won't this raise an `AttributeError`? `_strategy` is not defined anywhere.

Comment (sasha0) [inline src/oscar/apps/search/search_indexes.py]: Fixed.

**Pull request merged**